manager/src/main/java/org/openremote/manager/mqtt/TeltonikaMQTTHandler.java

```diff
@@ -0,0 +1,627 @@
+
+        //The AssetStateDuration will have a startTime of the found state-change variable, and the endDate will be the previousValue's timestamp, that being the end of the state duration
+
+        if(previousValue.getTimestamp().isEmpty()){
+            getLogger().warning("previousValue's timestamp is empty!");
+            return Optional.empty();
+        }
+
+        Attribute<?> tripAttr = new Attribute<>("LastTripStartedAndEndedAt", ValueType.ASSET_STATE_DURATION, new AssetStateDuration(
+                new Timestamp(StateChangeAssetDatapoint.getTimestamp()),
+                new Timestamp(previousValue.getTimestamp().get())
+        ));
+
+        tripAttr.setTimestamp(StateChangeAssetDatapoint.getTimestamp());
+
+        return Optional.of(tripAttr);
+    }
+
+    private String getParameterFileString() {
+        try {
+            return Files.readString(DeviceParameterPath);
+        } catch (IOException e) {
+            getLogger().warning("Couldn't find FMC003.json, couldn't parse parameters");
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Updates the {@link Asset} passed, with the {@link AttributeMap} passed.
+     * TODO: I have a very big feeling that the way this is done is wrong.
+     *
+     * @param asset The asset to be updated.
+     * @param attributes The attributes to be upserted to the Attribute.
+     * @param topic The topic to which the MQTT payload was sent.
+     * @param connection The connection on which the payload was sent.
+     */
+    private void updateAsset(Asset<?> asset, AttributeMap attributes, Topic topic, RemotingConnection connection) {
+
+        Attribute<Date> packetTime = attributes.get(new AttributeDescriptor<Date>("lastContact", ValueType.DATE_AND_TIME)).get();
+
+        Long packetTimestamp = packetTime.getValue().get().getTime();
+
+        String imei = asset.getAttribute("IMEI").toString();
+
+        getLogger().info("Updating CarAsset with IMEI "+imei);
+        //OBD details: Prot:6,VIN:WVGZZZ1TZBW068095,TM:15,CNT:19,ST:DATA REQUESTING,P1:0xBE3EA813,P2:0xA005B011,P3:0xFED00400,P4:0x0,MIL:0,DTC:0,ID3,Hdr:7E8,Phy:0
+        //Find which attributes need to be updated and which attributes need to be just reminded of updating.
+
+        //I'm not sure why this needs these specific headers.
+        Map<String, Object> headers = prepareHeaders(topicRealm(topic), connection);
+
+        AttributeMap nonExistingAttributes = new AttributeMap();
+
+        attributes.forEach( attribute ->  {
+            //Attribute exists, needs to be updated
+            if(asset.getAttributes().containsKey(attribute.getName())){
+                AttributeEvent attributeEvent = new AttributeEvent(asset.getId(), attribute.getName(), attribute.getValue(), packetTimestamp);
+                LOG.finer("Publishing to client inbound queue: " + attributeEvent);
+                messageBrokerService.getFluentProducerTemplate()
```

Comment:
You can use `AssetProcessingService.sendAttributeEvent()`

Reply:
Should I use `AttributeEvent.Source#SENSOR` as the `Source` in the overload of that function? Since it sort of is a sensor update, just handled by the MQTT Handler.